### PR TITLE
Added protocol plugins - continuation of #958

### PR DESCRIPTION
**Pull request by jesper-friis**
# Description
Continuation of PR #958, but with PR #939 merged in.

Added protocol plugins to separate protocol from storage driver. The protocol plugins are currently only implemented in Python.

Currently 3 protocols plugins has been added: file, http and sftp.

---

Suggestions from review:
- [x] Document that `dlite.Instance.from_url()` now support the following forms:
      - driver://netloc/path?options#id  # old form
      - protocol+driver://netloc/path?options#id
      - protocol://netloc/path?driver=driver;options#id  
- [x] Add test for file protocol with a file
- [x] Add test for file protocol with a directory
- [ ] Add test for sftp protocol against sftp server on localhost
- [x] Document that it is possible for a plugin to only implement the `from_bytes()` or `to_bytes()` class methods

---

## Type of change
- [ ] Bug fix & code cleanup
- [x] New feature
- [ ] Documentation update
- [ ] Test update

## Checklist for the reviewer
This checklist should be used as a help for the reviewer.

- [ ] Is the change limited to one issue?
- [ ] Does this PR close the issue?
- [ ] Is the code easy to read and understand?
- [ ] Do all new feature have an accompanying new test?
- [ ] Has the documentation been updated as necessary?

### CODE REVIEW

File: doc/user_guide/protocol_plugins.md [hunk 1]
```diff
@@ -0,0 +1,47 @@
+Protocol plugins
+================
+Protocol plugins is a new feature of DLite that allow a clear separation between serialising/parsing DLite instances to/from external data representations and transferring data to/from external data resources.
+This is illustrated in the figure below.
```
Comment (francescalb) [inline doc/user_guide/protocol_plugins.md:4]:
```suggestion
This is illustrated in the Figure below.
```

File: doc/user_guide/protocol_plugins.md [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    >>> url = "https://raw.githubusercontent.com/SINTEF/dlite/refs/heads/master/storages/python/tests-python/input/test_meta.yaml"
+    >>> dlite.Instance.load(protocol="http", driver="yaml", location=url)
+
+```
+
+Directly access to protocols
+----------------------------
+Protocol plugins can also be accessed in Python via the `Protocol` class.
+
+This example first save the data `b"hello world"` to the file `hello.txt` and then read load it back again:
```
Comment (francescalb) [inline doc/user_guide/protocol_plugins.md:29]:
```suggestion
In the example below the data `b"hello world"`  is first saved to the file `hello.txt` and then loaded back again:
```

File: doc/user_guide/storage_plugins.md [hunk 3]
```diff
@@ -137,51 +189,34 @@ In fact, even though `blob1` and `inst1` are different python objects, they shar
...
+- **open()**: Open and initiate the storage.  Required if any of load(), save(), delete(), query(), flush(), close() are defined.
+- **load()**: Load an instance from the storage and return it.
+- **save()**: Save an instance to the storage.
+- **delete()**: Delete and an instance from the storage.
+- **query()**: Query the storage for instance UUIDs.
+- **flush()**: Flushed cached data to the storage.
+- **close()**: Close the storage.
+- **from_bytes()**: Class method that loads an instance from a buffer.
+- **to_bytes()**: Class method that saves an instance to a buffer.
+
```
Comment (francescalb) [inline doc/user_guide/storage_plugins.md:206]:
```suggestion

A plugin can be defined with only the `from_bytes` and `to_bytes` methods. This has the advantage that only serialisation/deserialisation of the data and datamodels is considered in the actual plugin. In such cases Instances must
be instantiated with the dlite.Instance.load method, and the [protocol plugin] must be given.

```

**Comment by jesper-friis** [comment]
> Good, I am still missing a bit clearer description of which methods that must be implemented in a plugin and under which conditions.. See my suggestion. Something like this could (should) be added to the template, I think. Otherwise it seem OK.

Good suggestion. I committed it.

**Pull request merged**